types/src/validator_verifier.rs

```diff
@@ -159,54 +170,106 @@ impl ValidatorVerifier {
+            );
         }
-        Ok(())
+
+        let aggregated_key = PublicKey::aggregate(pub_keys_to_agg)
+            .map_err(|_| VerifyError::FailedToAggregatePubKey)?;
+        Ok((
+            MultiSignature::new(validator_bitmask, Some(aggregated_sig)),
+            aggregated_key,
+        ))
     }
 
-    /// This function will try batch signature verification and falls back to normal
-    /// iterated verification if batching fails.
-    pub fn batch_verify_aggregated_signatures<T: CryptoHash + Serialize>(
+    pub fn aggregate_and_verify_multi_signature<T: CryptoHash + Serialize>(
         &self,
+        partial_signatures: &PartialSignatures,
         message: &T,
-        aggregated_signature: &BTreeMap<AccountAddress, bls12381::Signature>,
+    ) -> Result<MultiSignature, VerifyError> {
+        let (aggregated_sig, aggregated_key) =
+            self.aggregate_multi_signature(partial_signatures)?;
+        // Verify the multi-signature
+        aggregated_sig
+            .multi_sig()
+            .as_ref()
+            .expect("Failed to get multi signature")
+            .verify(message, &aggregated_key)
+            .map_err(|_| VerifyError::FailedToVerifyMultiSignature)?;
+        Ok(aggregated_sig)
+    }
+
+    /// This function will successfully return when at least quorum_size signatures of known authors
+    /// are successfully verified. It creates an aggregated public key using the voter bitmask passed
+    /// in the multi-signature and verifies the message passed in the multi-signature using the aggregated
+    /// public key.
+    ///
+    pub fn verify_multi_signatures<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+        multi_signature: &MultiSignature,
     ) -> std::result::Result<(), VerifyError> {
-        self.check_num_of_signatures(aggregated_signature)?;
-        self.check_voting_power(aggregated_signature.keys())?;
-        let keys_and_signatures: Vec<(bls12381::PublicKey, bls12381::Signature)> =
-            aggregated_signature
-                .iter()
-                .flat_map(|(address, signature)| {
-                    let sig = signature.clone();
-                    self.get_public_key(address).map(|pub_key| (pub_key, sig))
-                })
-                .collect();
-        // Fallback is required to identify the source of the problem if batching fails.
-        if bls12381::Signature::batch_verify(message, keys_and_signatures).is_err() {
-            self.verify_aggregated_struct_signature(message, aggregated_signature)?
-        }
+        self.check_num_of_signatures(multi_signature)?;
+        self.check_voting_power(
+            multi_signature
+                .get_voter_addresses(&self.get_ordered_account_addresses_iter().collect_vec())
+                .iter(),
+        )?;
+        // Verify the optimistically aggregated signature.
+        let pub_keys_to_agg: Vec<&PublicKey> = multi_signature
+            .get_voters_bitmap()
```

Comment:
we already get voter_addresses from the check_voting_power, doesn't need to go through it again. just map to the self.address_to_validator_info seems cleaner